### PR TITLE
[SPARK-34731][CORE] Avoid ConcurrentModificationException when redacting properties in EventLoggingListener

### PR DESCRIPTION
**Pull request by bersprockets**
### What changes were proposed in this pull request?

Change DAGScheduler to pass a clone of the Properties object, rather than the original object, to the SparkListenerJobStart event.

### Why are the changes needed?

 DAGScheduler might modify the Properties object (e.g., in addPySparkConfigsToProperties) after firing off the SparkListenerJobStart event. Since the handler for that event (onJobStart in EventLoggingListener) will iterate over the elements of the Property object, this sometimes results in a ConcurrentModificationException.

This can be demonstrated using these steps:
```
$ bin/spark-shell --conf spark.ui.showConsoleProgress=false \
--conf spark.executor.cores=1 --driver-memory 4g --conf \
"spark.ui.showConsoleProgress=false" \
--conf spark.eventLog.enabled=true \
--conf spark.eventLog.dir=/tmp/spark-events
...
scala> (0 to 500).foreach { i =>
     |   val df = spark.range(0, 20000).toDF("a")
     |   df.filter("a > 12").count
     | }
21/03/12 18:16:44 ERROR AsyncEventQueue: Listener EventLoggingListener threw an exception
java.util.ConcurrentModificationException
	at java.util.Hashtable$Enumerator.next(Hashtable.java:1387)
```

I've not actually seen a ConcurrentModificationException in onStageSubmitted, only in onJobStart. However, they both iterate over the Properties object, so for safety's sake I pass a clone to SparkListenerStageSubmitted as well.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
By repeatedly running the reproduction steps from above.

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
add to whitelist

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Similar issue was reported from SPARK-32027 - probably the same occurrence or not.

As I commented in SPARK-32027, I think the fundamental problem is that mutability on event object (including all fields in event object of course), which would be problematic via 1) co-use of event object between listener"s" 2) co-use of specific objects between even object and main logic. But I also couldn't evaluate the "cost" of fixing fundamental problem altogether, so fixing one by one once we figure out would make sense.

Thanks for submitting a PR! I'll take a look today.

File: core/src/main/scala/org/apache/spark/util/Utils.scala [hunk 1]
```diff
@@ -3003,6 +3003,12 @@ private[spark] object Utils extends Logging {
     resultProps
   }
 
+  /** Create a new properties object with the same values as `props` while handling null case */
```
Comment (HeartSaVioR) [inline core/src/main/scala/org/apache/spark/util/Utils.scala:3006]: I'd rather say we'd be better to fix `cloneProperties` to handle null for props instead: it just throws NPE now.

**Review by HeartSaVioR — COMMENTED** [review]
Looks OK to me, only a comment.

**Comment by mridulm** [comment]
This should go away when we drop support for jdk 8 ... from jdk 9, it is stored in a `ConcurrentHashMap`

**Comment by HeartSaVioR** [comment]
Good to know, thanks! 

Btw, we haven't considered anything on dropping JDK 8 as well if I understand correctly, so I'd still want to see the fix at least on cases which are figured out. Furthermore, that's more likely "implementation details" rather than "interface/method contract", so  I'm not 100% sure that's something we can rely on. (Is there javadoc change or guide doc change on Java 9 Properties?)

File: core/src/main/scala/org/apache/spark/util/Utils.scala (hunk 1 above)
Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/util/Utils.scala]: Thanks for looking at this. I agree, I don't know why I didn't do that in the first place.

**Review by HeartSaVioR — APPROVED** [review]
+1

I don't feel I should interpret @mridulm 's comment as "let's drop JDK 8 instead of this fix". For me, it sounds more likely providing better possibility, as this PR doesn't guarantee we fix every spots it could bring concurrent issue. Please let me know if it's not the case.

**Comment by bersprockets** [comment]
My commit did not restart tests, which is a little weird (or things have changed in the way that tests runs are indicated since I last made a PR).

**Comment by HeartSaVioR** [comment]
ok to test

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
Don't worry too much if Github Action passes. If I'm not missing here, GA passed is considered as build passed nowadays.

**Comment by HeartSaVioR** [comment]
I'll wait a day to see whether there's further comment (including @mridulm feedback on my interpretation) and merge if there's no outstanding comment until then.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HeartSaVioR** [comment]
I guess Jenkins is not healthy. I'll trigger Github Action and merge if it passes.

**Comment by HyukjinKwon** [comment]
Scala 2.13 is fixed in https://github.com/apache/spark/commit/c5cadfefdf9b4c6135355b49366fd9e9d1e3fcd0. Let me just merge this in @HeartSaVioR.

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.1.

**Comment by HeartSaVioR** [comment]
Thanks @HyukjinKwon for taking this over! I missed the right time.

**Comment by bersprockets** [comment]
Thanks @HeartSaVioR @HyukjinKwon !

**Pull request closed without merging**